### ISSUE TITLE
OSDs get into pending state with topologySpreadConstraints

### ISSUE THREAD

**Issue by kesavanvt** [rook/rook#5932]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
with topologySpreadConstraints and storageClassDeviceSets count as 6, some of the OSDs gets into pending state

I believe its something related to #5586, as the osds and osd prepare job shares the same placement config

**Expected behavior:**
OSD pods should be evenly distributed based on the topologySpreadConstraints provided (i.e 2 OSDs per node here)

**How to reproduce it (minimal and precise):**
On an AWS Openshift Cluster

- Apply common.yaml, operator-openshift.yaml
- Apply cluster-on-pvc.yaml as below

**File(s) to submit**:
* cluster-on-pvc.yaml
``` yaml
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  dataDirHostPath: /var/lib/rook
  mon:
    count: 3
    allowMultiplePerNode: false
    volumeClaimTemplate:
      spec:
        storageClassName: gp2
        resources:
          requests:
            storage: 10Gi
  cephVersion:
    image: ceph/ceph:v15.2.4
    allowUnsupported: false
  skipUpgradeChecks: false
  continueUpgradeAfterChecksEvenIfNotHealthy: false
  mgr:
    modules:
    - name: pg_autoscaler
      enabled: true
  dashboard:
    enabled: true
    ssl: true
  crashCollector:
    disable: false
  storage:
    storageClassDeviceSets:
    - name: set1
      count: 6
      portable: true
      tuneDeviceClass: true
      placement:
        topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: DoNotSchedule
          labelSelector:
            matchExpressions:
            - key: app
              operator: In
              values:
              - rook-ceph-osd
              - rook-ceph-osd-prepare
      resources:
      volumeClaimTemplates:
      - metadata:
          name: data
        spec:
          resources:
            requests:
              storage: 10Gi
          storageClassName: gp2
          volumeMode: Block
          accessModes:
            - ReadWriteOnce
  disruptionManagement:
    managePodBudgets: false
    osdMaintenanceTimeout: 30
    manageMachineDisruptionBudgets: false
    machineDisruptionBudgetNamespace: openshift-machine-api
```

* When both `rook-ceph-osd` and `rook-ceph-osd-prepare` provided as labelselector for topologySpreadConstraints
```
oc get pods -l 'app in (rook-ceph-osd,rook-ceph-osd-prepare)' -o wide 
NAME                                            READY   STATUS      RESTARTS   AGE   IP            NODE                                         NOMINATED NODE   READINESS GATES
rook-ceph-osd-0-977bfcf48-v8kp8                 0/1     Pending     0          22m   <none>        <none>                                       <none>           <none>
rook-ceph-osd-1-7bb84dbb49-6lm86                1/1     Running     0          22m   10.128.2.22   ip-10-0-206-137.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-2-d56cbd9fb-kl894                 1/1     Running     0          21m   10.131.0.21   ip-10-0-186-216.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-3-69b98dd94c-rz8pd                1/1     Running     0          21m   10.129.2.17   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-0-hvtk2-89jgq   0/1     Completed   0          22m   10.131.0.19   ip-10-0-186-216.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-1-9mqqw-l48ld   0/1     Completed   0          22m   10.128.2.21   ip-10-0-206-137.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-2-j8nrr-dmf2b   0/1     Completed   0          22m   10.129.2.15   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-3-sjbwt-stmnb   0/1     Completed   0          22m   10.129.2.16   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-4-vgcl8-86sc9   0/1     Pending     0          22m   <none>        <none>                                       <none>           <none>
rook-ceph-osd-prepare-set1-data-5-jxwf9-zk2lw   0/1     Pending     0          22m   <none>        <none>                                       <none>           <none>
```
* When `rook-ceph-osd` is provided as labelselector for topologySpreadConstraints
```
oc get pods -l 'app in (rook-ceph-osd,rook-ceph-osd-prepare)' -o wide
NAME                                            READY   STATUS      RESTARTS   AGE     IP            NODE                                         NOMINATED NODE   READINESS GATES
rook-ceph-osd-0-74b9455f8f-f2q7w                1/1     Running     0          2m30s   10.128.2.33   ip-10-0-206-137.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-1-7766f8c69b-r5pwn                0/1     Pending     0          2m27s   <none>        <none>                                       <none>           <none>
rook-ceph-osd-2-7d7bc6ccf6-jf8zr                1/1     Running     0          2m22s   10.129.2.29   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-3-64c9b9f7b-bkhq4                 0/1     Pending     0          2m20s   <none>        <none>                                       <none>           <none>
rook-ceph-osd-4-86568dc8b9-2kwk8                0/1     Pending     0          2m17s   <none>        <none>                                       <none>           <none>
rook-ceph-osd-5-5948f5844d-mx9g8                0/1     Pending     0          2m15s   <none>        <none>                                       <none>           <none>
rook-ceph-osd-prepare-set1-data-0-5x9fs-7xv66   0/1     Completed   0          2m51s   10.129.2.26   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-1-9tgvl-s7jrn   0/1     Completed   0          2m50s   10.128.2.30   ip-10-0-206-137.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-2-w6spn-4n66m   0/1     Completed   0          2m50s   10.129.2.27   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-3-r5vph-cjznr   0/1     Completed   0          2m49s   10.129.2.28   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-4-s4fpc-h9bfb   0/1     Completed   0          2m49s   10.128.2.31   ip-10-0-206-137.us-east-2.compute.internal   <none>           <none>
rook-ceph-osd-prepare-set1-data-5-kv6zg-57kz9   0/1     Completed   0          2m48s   10.129.2.25   ip-10-0-139-123.us-east-2.compute.internal   <none>           <none>
```
**Workarounds:**
The OSDs tends to be running and evenly distributed by the following workarounds

workaround 1

- Making whenUnsatisfiable as `ScheduleAnyway` and with `rook-ceph-osd `and `rook-ceph-osd-prepare` as labelSelector.
I am not sure if its guaranteed to provide even spread for variable counts since it has ScheduleAnyway. 

workaround 2
- With only `rook-ceph-osd-prepare` as labelSelector and whenUnsatisfiable as `DoNotSchedule`

looking for the suggestion on the ideal way for consistent even spread

**Environment**:
* Cloud provider or hardware configuration: aws
* Rook version (use `rook version` inside of a Rook Pod): v1.3.0
* Storage backend version (e.g. for ceph do `ceph -v`): v15.2.4
* Kubernetes version (use `kubectl version`): v1.18.3+3107688
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): OpenShift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): HEALTH_OK

**Comment by travisn** [comment]
We really need to apply separate spread constraints for the job vs the deployment. The osd prepare jobs should only be compared to each other, and the osd daemons should only look at other osd daemons. If we try and mix these, we won't end up with the desired results. So we need a change in rook that will separate how it applies the constraints.

**Comment by travisn** [comment]
@satoru-takeuchi Have you seen this issue when testing topology spread constraints?

**Comment by satoru-takeuchi** [comment]
@travisn No I've not. I'll try to reproduce this problem in my test environment next week.

**Comment by satoru-takeuchi** [comment]
@travisn 

> We really need to apply separate spread constraints for the job vs the deployment. The osd prepare jobs should only be
> compared to each other, and the osd daemons should only look at other osd daemons. If we try and mix these, we won't
> end up with the desired results. So we need a change in rook that will separate how it applies the constraints.

I consider it's only applied to "portable: true" case. Is it correct? You said both kind of pod should have the same placement before.

https://github.com/rook/rook/issues/4582#issuecomment-573133915
> I believe the operator should be setting the same placement rules on the osd prepare pods as on the osd daemons.

**Comment by jarrpa** [comment]
I would say it should be applied in all cases. When `portable` is false, the PV should have NodeAffinity associated with it to inform the scheduling of the Pod once the PVC it's going to use binds to said PV. Even then, when you're increasing the number of OSDs, you don't want the new Jobs blocking the new OSD Pods.

**Comment by travisn** [comment]
@satoru-takeuchi Agreed that this concern only applies to `portable: true`. When `portable: false`, the osd prepare jobs will have the placement settings applied, but then the only setting on the osd daemon deployments is a node-selector, which is the hostname where the prepare job ran. If it's not portable, we can't allow the scheduler to pick a different node. 
@jarrpa If it's not portable, agreed that we would expect the PV to have node affinity already. But if they want the non-portable behavior even when the underlying PVs are portable, we still enforce it with the osd daemon deployment node affinity.

**Comment by satoru-takeuchi** [comment]
@travisn @jarrpa I confirmed this problem wouldn't happen when `portable: false`. At least, my team has used topologySpreadConstraints with `portable: false` and there hasn't been any problem in even-osd-pod-spreading.

In addition, I also consider the best solution is separating placement between osd prepare pods and osd pods.

Here is my opinion about workarounds.

```
workaround 1

Making whenUnsatisfiable as ScheduleAnyway and with rook-ceph-osd and rook-ceph-osd-prepare as labelSelector.
I am not sure if its guaranteed to provide even spread for variable counts since it has ScheduleAnyway.
workaround 2

With only rook-ceph-osd-prepare as labelSelector and whenUnsatisfiable as DoNotSchedule
```

I consider only workaround1 works correctly.

First, `scheduleAnyway` doesn't work well by default. If you want to use workaround1, it works fine only if we use custom scheduling policy to increasing the weight of even pod spreading in scoring nodes. My team also uses `scheduleAnyway`.

For more detail, please see the following issue/PR.

https://github.com/rook/rook/issues/5536
https://github.com/rook/rook/pull/5630

As for workaround1, I guess it doesn't work fine. For example, if you'll add another OSDs one by one after creating the initial six osd pods, the other osd prepare pods, corresponding to the running osd pods, are already finished. Then topologySpreadConstraints do nothing because this feature only counts on the running pods.

In any case, I suggest `scheduleAnyway` because it's really difficult to appropriate `maxSkew`.

**Comment by travisn** [comment]
@kesavanvt After analyzing this issue more, I believe we need to go with:
1. Set both `rook-ceph-osd` and `rook-ceph-osd-prepare` as labelselectors for the topologySpreadConstraints
1. Set `whenUnsatisfiable: ScheduleAnyway`

This is based on the assumption from the [k8s docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#spread-constraints-for-pods) that :
```
ScheduleAnyway tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.
```

The reason that we must specify `ScheduleAnyway` is that for every OSD there are **two pods**: a prepare pod and the daemon pod. If both of these attempt to be scheduled before the prepare pod for another OSD is scheduled, the skew would be 2 and would not be schedulable. After this happens, all the scheduling of OSDs is blocked. But if we allow the osd to `ScheduleAnyway`, the OSDs will be unbalanced in the short term, but as soon as the next osd is scheduled, they will even out and reduce the skew again.

In other words, I believe we will get the desired behavior even though the skew will temporarily be two sometimes. There may be some corner cases where nodes are not schedulable for other reasons, but let's start with these assumptions and see what improvements are needed in the future. 

@kesavanvt In your two examples above, the first example only placed OSDs on three nodes and the second example only placed OSDs on two nodes. It doesn't appear in your test that you had six nodes that were available for OSDs, so the nodes must need to be labeled differently.

**Comment by satoru-takeuchi** [comment]
@travisn @kesavanvt 

> ScheduleAnyway tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.

k8s scheduler guys admit the description is wrong.

https://github.com/kubernetes/kubernetes/issues/91234#issuecomment-630989950
> I agree that the ScheduleAnyway semantics are confusing, I wish we can change the whenUnsatisfiable stanza, but I think it is too late, and so the best we can do is clarify in the documentation.

The suggested workaround is `ScheduleAnyway` + tweaking the scoring weight of even-pod-spreading in Pod scheduling. Without doing so, the weight of even-pod-spreading is very low. Then, If there is a heavy workload in a node with disks, these disks won't be used as expected.

FYI, Here is my team's configuration. Here we use k8s 1.17 and can set only one scoring policy.

https://github.com/cybozu-go/neco/blob/master/etc/cke-template.yml#L30-L90

After k8s 1.18, we can use the scheduling profile feature (alpha). Then we can set multiple scheduling policies and we can set the appropriate policy, like the one in the above link, with the following feature.

https://github.com/rook/rook/pull/5630
